fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java

```diff
@@ -0,0 +1,464 @@
+            }
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Boolean visitLogicalTopN(OptExpression optExpression, Void context) {
+            if (!allowSort) {
+                return false;
+            }
+            allowSort = false;
+            allowLimit = false;
+
+            LogicalTopNOperator topNOp = optExpression.getOp().cast();
+            List<Ordering> orderings = topNOp.getOrderByElements();
+            if (!orderings.stream().allMatch(Ordering::isAscending)) {
+                return false;
+            }
+            orderByColumns = orderings.stream()
+                    .map(Ordering::getColumnRef)
+                    .map(ColumnRefOperator::getName)
+                    .map(String::toLowerCase)
+                    .collect(Collectors.toList());
+            if (topNOp.getOffset() != Operator.DEFAULT_OFFSET) {
+                return false;
+            }
+            if (topNOp.getLimit() != Operator.DEFAULT_LIMIT) {
+                limit = topNOp.getLimit();
+            }
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Boolean visitLogicalLimit(OptExpression optExpression, Void context) {
+            if (!allowLimit) {
+                return false;
+            }
+            allowLimit = false;
+
+            LogicalLimitOperator limitOp = optExpression.getOp().cast();
+            if (limitOp.getOffset() != 0) {
+                return false;
+            }
+            limit = limitOp.getLimit();
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Boolean visitLogicalTableScan(OptExpression optExpression, Void context) {
+            return createLogicalPlanChecker(optExpression, allowFilter, allowLimit, allowProject,
+                    allowSort, predicate, orderByColumns, limit).visitLogicalTableScan(optExpression, context);
+        }
+
+        protected static boolean isPointScan(Table table, List<String> keyColumns, List<ScalarOperator> conjuncts) {
+            Map<String, PartitionColumnFilter> filters = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+            filters.putAll(ColumnFilterConverter.convertColumnFilter(conjuncts, table));
+            if (keyColumns == null || keyColumns.isEmpty()) {
+                return false;
+            }
+            long cardinality = 1;
+            for (String keyColumn : keyColumns) {
+                if (!filters.containsKey(keyColumn)) {
+                    return false;
+                }
+                PartitionColumnFilter filter = filters.get(keyColumn);
+                if (filter.getInPredicateLiterals() != null) {
+                    cardinality *= filter.getInPredicateLiterals().size();
+                    if (cardinality > MAX_RETURN_ROWS) {
+                        return false;
+                    }
+                } else if (!filter.isPoint()) {
```

Comment:
can between predicate and binary predicate be used for short-circuit read? some of them can be extended into in-predicates